runtime/lua/vim/lsp/handlers.lua

```diff
@@ -593,8 +593,19 @@ local function make_type_hierarchy_handler()
         col = col + 1,
       })
     end
-    vim.fn.setqflist({}, ' ', { title = 'LSP type hierarchy', items = items })
-    api.nvim_command('botright copen')
+    config = config or {}
+    local title = 'LSP type hierarchy'
+
+    if config.loclist then
+      vim.fn.setloclist(0, {}, ' ', { title = title, items = items, context = ctx })
+      api.nvim_command('lopen')
+    elseif config.on_list then
+      assert(type(config.on_list) == 'function', 'on_list is not a function')
+      config.on_list({ title = title, items = items, context = ctx })
+    else
+      vim.fn.setqflist({}, ' ', { title = title, items = items, context = ctx })
+      api.nvim_command('botright copen')
+    end
```

Comment:
You can even make the entire table a shared variable:
```suggestion
    config = config or {}
    local list = { title = 'LSP type hierarchy', items = items, context = ctx }

    if config.loclist then
      vim.fn.setloclist(0, {}, ' ', list)
      api.nvim_command('lopen')
    elseif config.on_list then
      assert(type(config.on_list) == 'function', 'on_list is not a function')
      config.on_list(list)
    else
      vim.fn.setqflist({}, ' ', list)
      api.nvim_command('botright copen')
    end
```